lib/vrl/stdlib/src/get_enrichment_table_record.rs

```diff
@@ -0,0 +1,189 @@
+use vrl::{enrichment::Condition, prelude::*};
+
+#[derive(Clone, Copy, Debug)]
+pub struct GetEnrichmentTableRecord;
+impl Function for GetEnrichmentTableRecord {
+    fn identifier(&self) -> &'static str {
+        "get_enrichment_table_record"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "table",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "condition",
+                kind: kind::OBJECT,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[]
+    }
+
+    fn compile(&self, state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let tables = state
+            .get_enrichment_tables()
+            .as_ref()
+            .map(|tables| {
+                tables
+                    .table_ids()
+                    .into_iter()
+                    .map(Value::from)
+                    .collect::<Vec<_>>()
+            })
+            .unwrap_or_else(Vec::new);
+
+        let table = arguments.required_enum("table", &tables)?.to_string();
+        let condition = arguments.required_object("condition")?;
+
+        Ok(Box::new(FindTableRowFn { table, condition }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GetEnrichmentTableRecordFn {
+    table: String,
+    condition: BTreeMap<String, expression::Expr>,
+}
+
+impl Expression for GetEnrichmentTableRecordFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let condition = self
+            .condition
+            .iter()
+            .map(|(key, value)| {
+                Ok(Condition::Equals {
+                    field: key,
+                    value: value.resolve(ctx)?.try_bytes_utf8_lossy()?.to_string(),
+                })
+            })
+            .collect::<Result<Vec<Condition>>>()?;
+
+        let tables = ctx
+            .get_enrichment_tables()
+            .ok_or("enrichment tables not loaded")?;
+
+        match tables.find_table_row(&self.table, &condition)? {
+            None => Err("data not found".into()),
+            Some(data) => Ok(Value::Object(data)),
+        }
+    }
+
+    fn update_state(
+        &self,
+        state: &mut state::Compiler,
+    ) -> std::result::Result<(), ExpressionError> {
+        match state.get_enrichment_tables_mut() {
+            Some(ref mut table) => {
+                let fields = self
+                    .condition
+                    .iter()
+                    .map(|(field, _)| field.as_ref())
+                    .collect::<Vec<_>>();
+                table.add_index(&self.table, &fields)?;
+                Ok(())
+            }
+            // We shouldn't reach this point since the type checker will ensure the table exists before this function is called.
+            None => Err("enrichment tables aren't loaded".into()),
```

Comment:
Do we want to have an `unreachable` here instead? We do that in other places where we want a panic to occur if we have an error in our compiler implementation.